ISPN-5591 Simple local cache without interceptor stack

rvansa opened the pull request:
* the cache is returned as default for local cache when transactions, persistence
  indexing, storeAsBinary or jmx statistics is not used. As e.g. map reduce or
  distributed executors are not a part of configuration, to run these
  (unsupported by SimpleLocalCacheImpl) setting allow-simple="false"
  is required as a new configuration attribute.
* also passing metadata to certain methods of CacheNotifier as param

rvansa commented:
These are the results I got on server machine:
```
MapBenchmark.testGetChm             thrpt   25  128354135.649 ± 2178755.959  ops/s
MapBenchmark.testGetDataContainer   thrpt   25  112883338.149 ±  190628.202  ops/s
MapBenchmark.testGetSimpleCache     thrpt   25   86969897.322 ±  738935.544  ops/s
MapBenchmark.testGetCache           thrpt   25   17280018.799 ±  361910.033  ops/s
MapBenchmark.testPutChm             thrpt   25   33980088.875 ±   28487.932  ops/s
MapBenchmark.testPutDataContainer   thrpt   25   17670045.892 ±   27393.106  ops/s
MapBenchmark.testPutDataContainer2  thrpt   25   24293649.544 ±  694969.064  ops/s
MapBenchmark.testPutSimpleCache     thrpt   25   14044642.521 ±   14280.945  ops/s
MapBenchmark.testPutCache           thrpt   25    2267850.070 ±   44814.045  ops/s
```

The only difference in configuration is allow-simple="true|false". So it seems that for simple puts and gets the simple cache is 5x faster for gets and 6x faster for puts.

rvansa commented:
I'll try to add the statistics support, too - I'll see what perf impact that will have.

wburns commented on commons/src/main/java/org/infinispan/commons/equivalence/EntryEquivalence.java:
```diff
@@ -0,0 +1,62 @@
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class EntryEquivalence<K, V> implements Equivalence<Map.Entry<K, V>> {
+	private final Equivalence<K> keyEquivalence;
```
This class has tabs instead of 3 spaces.

wburns commented on commons/src/main/java/org/infinispan/commons/util/CloseableIteratorAdapter.java:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public final class CloseableIteratorAdapter<E> implements CloseableIterator<E> {
```
This class is already implemented https://github.com/infinispan/infinispan/blob/master/commons/src/main/java/org/infinispan/commons/util/IteratorAsCloseableIterator.java

You can tweak that one to allow \<? extends E\> though.  Although I am actually moving this in https://github.com/infinispan/infinispan/pull/3569/files#diff-98ed1727054c0e59fa45ca9079aa4d6eR39

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+ */
+@MBean(objectName = CacheImpl.OBJECT_NAME, description = "Component that represents a simplified cache instance.")
+public class SimpleLocalCacheImpl<K, V> implements AdvancedCache<K, V> {
+	private final static Log log = LogFactory.getLog(SimpleLocalCacheImpl.class);
```
This guy has tabs as well.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -52,6 +73,39 @@
+      StoreAsBinaryConfiguration storeAsBinary = configuration.storeAsBinary();
+      JMXStatisticsConfiguration jmxStatistics = configuration.jmxStatistics();
+      // TODO: eventually support transactional caches as well
+      if (clustering != null && clustering.cacheMode().isClustered()) {
```
These don't need to be a bunch of if elses since you are returning in the if.  It seems this would be cleaner if it was just a local variable assignment and then use a single appropriately for that variable.

```
ClusteringConfiguration clustering = configuration.clustering();
if (clustering != null && clustering.cacheMode().isClustered()) {
   return false;
}
TransactionConfiguration transaction = configuration.transaction();
if (transaction != null && transaction.transactionMode().isTransactional()) {
   return false;
}
... etc ...
```

rvansa replied:
I find the if-else syntax more compact, but if you prefer it that way, I won't object.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -92,4 +168,5 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
    public <T> T construct(Class<T> componentType) {
       throw new UnsupportedOperationException("Should never be invoked - this is a bootstrap factory.");
    }
+
```
Extra line not needed :)

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerStub.java:
```diff
@@ -0,0 +1,95 @@
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class PersistenceManagerStub implements PersistenceManager {
+	@Override
```
Weird indention

wburns commented on commons/src/main/java/org/infinispan/commons/equivalence/EntryEquivalence.java:
```diff
@@ -0,0 +1,62 @@
+	private final Equivalence<K> keyEquivalence;
+	private final Equivalence<V> valueEquivalence;
+
+	public EntryEquivalence(Equivalence<K> keyEquivalence, Equivalence<V> valueEquivalence) {
```
These should be \<? super K\> and \<? super V\>

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+
+	@Override
+	public boolean isEmpty() {
+		return getDataContainer().size() == 0;
```
This would be more performant if we used the iterator and found the first non expired entry.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+						return oldEntry;
+					});
+				} else {
+					++size;
```
These can have int overflows.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+				if (entry.isExpired(now)) {
+					final long NOW = now;
+					dataContainer.compute(entry.getKey(), (key, oldEntry, factory) -> {
+						if (entry.isExpired(NOW)) return null;
```
Believe you want oldEntry.isExpired(NOW) otherwise a concurrent write may be lost.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+
+	@Override
+	public DataContainer<K, V> getDataContainer() {
+		if (dataContainer == null) {
```
It seems we should store the data container in a local variable before doing the null check this way a concurrent stop wouldn't give a NPE back possibly.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+	}
+
+	private void assertKeyNotNull(Object key) {
+		if (key == null) {
```
We can start using Objects static methods for this now.

wburns commented:
Also just wanted to make sure you saw the test failures here http://ci.infinispan.org/viewLog.html?buildId=28060&buildTypeId=bt9&guest=1

danberindei commented on core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java:
```diff
@@ -0,0 +1,1608 @@
+			CacheNotifier cacheNotifier,
+			TimeService timeService,
+			InternalEntryFactory entryFactory
+	) {
```
We don't usually put the closing bracket on its own line (or the first parameter on a new line).

danberindei commented:
@rvansa Could you try your benchmark on the existing impl after removing the `keyEquivalence.equals(key, this.key)` check from `SingleKeyNonTxInvocationContext.lookupEntry(key)` and `SingleKeyNonTxInvocationContext.lookupEntry(key)`? They're not that useful, I think, and I was wondering if it's worth removing them to gain a bit of speed.

rvansa commented:
@wburns Thanks, for the comments, I'll go fix the testsuite... It seemed to me that I have a clean build, but I had a wrong look, probably.
Wrt code style: it's a pity we don't have checkstyle script for Infinispan... have to switch back from tabs (used in Hibernate) to chars in my IDE...

rvansa commented:
@danberindei: increased the number of iterations to reduce margin of error.
Without keyEquivalence checks:
```
MapBenchmark.testGetCache  thrpt   50  17818258.540 ± 112743.078  ops/s
```
With keyEquivalence checks:
```
MapBenchmark.testGetCache  thrpt   50  17404864.177 ± 115976.438  ops/s
```
That's about 2.3% speedup, not much.

danberindei commented:
@rvansa Thanks for running the benchmark, I was expecting a little more but I guess we don't call `lookupEntry()` that much. Maybe it would matter slightly more for writes, which need an additional equality check for `addLockedKey(key)`, but it's pretty sure that's not the main culprit.

rvansa commented:
@wburns @danberindei Updated, see commit comments. I hope the testsuite will be green this time (actually only a few tests are executed for SimpleCacheImpl now, since I haven't forked all the applicable tests for that - should I?)

rvansa commented:
Seems that the test failures are not related in any way.

danberindei commented on commons/src/main/java/org/infinispan/commons/util/ByRef.java:
```diff
@@ -25,4 +25,71 @@ public T get() {
+      }
+   }
+
+   public static class Pair<T1, T2> {
```
Doesn't really seem related to `ByRef`...

rvansa replied:
The idea is that these are two values that are returned by reference (in order to reduce the number of arguments passed to the function and allocations required), while if I had a generic pair, I would prefer to have it immutable.

danberindei replied:
Yeah, but it could just as well be a separate class called `ImmutablePair`

danberindei commented on core/src/main/java/org/infinispan/cache/impl/StatsCollectingCache.java:
```diff
@@ -0,0 +1,356 @@
+   public CloseableIteratorSet<Entry<K, V>> entrySet() {
+      return new EntrySet() {
+         @Override
+         protected int addAllInternal(Collection<? extends Entry<K, V>> c) {
```
`entrySet()` shouldn't support `add()` and `addAll()`

rvansa replied:
I'd guess so as well, and I wouldn't bother implementing it but CacheImpl implements that through https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/commands/read/EntrySetCommand.java#L105 and there are tests that check this behaviour.

wburns replied:
Good catch, I didn't notice that [1] defines this already.  I logged an issue [2] to remove support in other places.

[1] https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--
[2] https://issues.jboss.org/browse/ISPN-5621

danberindei replied:
I guess @wburns was a bit overzealous with the implementation :) The `Map.entrySet()` javadoc says `add()` isn't supported, so I think it would be best to remove it.

wburns replied:
@danberindei Isn't it fun to add stuff we never had before? :)

danberindei commented on core/src/main/java/org/infinispan/configuration/cache/CacheMode.java:
```diff
@@ -5,6 +5,11 @@
  */
 public enum CacheMode {
    /**
+    * Data is not replicated and certain features are not possible
```
I think this would be a good place to document what features are possible...

OTOH, I'm not sure adding a new cache mode is really the best way to expose this.

rvansa replied:
The cache cannot be injected transparently, and from configuration perspective there would be many prohibited configurations when you'd set the simple-cache="true" attribute. As this is not an orthogonal configuration option, I've decided to have that as new cache mode.
Can you support your opinion with any arguments we could discuss?

rvansa replied:
I had XML configuration in mind - the checks against prohibited configuration option are needed for programmatic configuration anyway.
And I agree that I could elaborate about the features a bit more here, though the documentation of other modes is not verbose neither.

danberindei replied:
@rvansa My main reason to not make SIMPLE a cache mode is that the current definition of what a SIMPLE cache seems very closely tied to the needs of Hibernate. For someone else, I'm not sure the trade-offs will be the same, for instance why support eviction but not persistence? So I would say at the very least it needs a little bake-in time before we expose it on the same level as LOCAL.

rvansa replied:
It's not about Hibernate needs, it's about rewrapping `DataContainer` as cache. So it's more like 'what we can get without too much effort from technical perspective'.
Btw. Hibernate does not need listeners, nor statistics, I've added these to fill in the contract as closely as possible.

danberindei replied:
I guess 'what we can get without too much effort' just seems a too fluid to me. Maybe at some point we'll be able to add an interceptor stack on the fly, and then we won't need any configuration for this...

@tristantarrant WDYT about making simple-cache a separate cache mode (and configuration element)?

danberindei commented on core/src/main/java/org/infinispan/configuration/parsing/Parser80.java:
```diff
@@ -1083,6 +1107,31 @@ protected void parseCacheElement(XMLExtendedStreamReader reader, Element element
+      }
+   }
+
+   protected void parseFullCacheElement(XMLExtendedStreamReader reader, Element element, ConfigurationBuilderHolder holder) throws XMLStreamException {
```
Better to invent new method names for the new mode...

rvansa replied:
These are shared argument for any cache. I am not very happy about the name 'full' cache either - any suggestions?

danberindei replied:
`parseSimpleCacheElement` for the other method?

rvansa replied:
simple cache is-a cache, full cache is-a cache (and local, dist and repl are full caches), but full-cache is-not-a simple cache, therefore it wouldn't be correct...
But ok, I can probably use `parseBaseCacheElement`...

danberindei commented on cdi/src/test/java/org/infinispan/cdi/test/event/CacheEventTest.java:
```diff
@@ -155,7 +155,7 @@ public void testFiringEntryModifiedEventWhenModifyingEntryInCache() throws Excep
    public void testFiringEntryInvalidatedWhenUsingCacheNotifier() throws Exception {
       //when
-      cache1Notifier.notifyCacheEntryInvalidated("pete", "Edinburgh", true, invocationContext, null);
+      cache1Notifier.notifyCacheEntryInvalidated("pete", "Edinburgh", null, true, invocationContext, null);
```
Are the notification changes related to the simple cache or could they be separated?

rvansa replied:
True, I could refactor that out. The problem was that notifier was reading some values from InvocationContext, while simple cache uses a dummy-one, so I've changed this to provide the values directly.

danberindei replied:
Ok, I was worried this might have been an actual bug in the current implementation.

danberindei commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -0,0 +1,1668 @@
+      Objects.requireNonNull(value, NULL_VALUES_NOT_SUPPORTED);
+      ByRef<InternalCacheEntry<K, V>> oldEntryRef = new ByRef<>(null);
+      boolean hasListeners = this.hasListeners;
+      // TODO: we will probably hit runtime error here if key is not of type K, but we can't even do a check here
```
`remove(Boolean.TRUE)` should do nothing in a `Cache<String, String>` (or `Map<String, String>`, for that matter).

rvansa replied:
but `remove(new Object() { public boolean equals(Object other) { return true; } });` should remove String (as long as hashcodes match), according to `Map` contract. Is `Cache` somehow narrowing that (haven't seen that in docs)?

danberindei replied:
Exactly, `Cache` isn't narrowing that, so throwing an exception at runtime isn't ok :)

Actually, I don't expect any `ClassCastException` here, because `K` is erased both in the `Cache` and in the `DataContainer`. But you should definitely have a test for it (and remove the TODO).

danberindei commented on commons/src/main/java/org/infinispan/commons/util/ByRef.java:
```diff
@@ -25,4 +25,71 @@ public T get() {
+      }
+   }
+
+   public static class Quadruple<T1, T2, T3, T4> {
```
Based on usage, this should probably be called `CacheEntryChange`.

danberindei commented on core/src/test/java/org/infinispan/api/APINonTxTest.java:
```diff
@@ -1,6 +1,7 @@
 package org.infinispan.api;
 
 import org.infinispan.commons.util.ObjectDuplicator;
+import org.infinispan.configuration.cache.CacheMode;
```
Not needed

danberindei commented on core/src/test/java/org/infinispan/api/AsyncAPITest.java:
```diff
@@ -14,21 +16,29 @@
 
+   @Override
+   protected void teardown() {
+      TestingUtil.replaceComponent(cacheManager, TimeService.class, new DefaultTimeService(), true);
```
The cache manager is being shut down anyway, do we really need to replace the time service?

rvansa replied:
I am not sure, but CacheNotifierTest goes that way (replacing the component back in tearDown) so I took that approach. I can't remember if I had any test failure without that, I could try...

danberindei replied:
Please do, I'm not sure why `CacheNotifierTest` does it but it shouldn't be necessary.

danberindei commented on core/src/test/java/org/infinispan/api/AsyncAPITest.java:
```diff
@@ -339,29 +349,52 @@ private void verifyEviction(final String key, final String expectedValue, final
+         };
+         assertTrue(expectedValue.equals(c.get(key)) || timeService.isTimeExpired(expectedEndTime));
+         while (!condition.isSatisfied() || !timeService.isTimeExpired(expectedEndTime)) {
+            timeService.advance(Math.max(1, expectedLifetime / 10));
```
Do you still need to repeat the loop 10 times? It should be enough to advance once to `expectedLifetime * 9 / 10`

danberindei commented on core/src/test/java/org/infinispan/notifications/cachelistener/CacheNotifierTest.java:
```diff
@@ -57,16 +63,26 @@ public void setUp() throws Exception {
       mockNotifier = mock(CacheNotifier.class);
       origNotifier = TestingUtil.replaceComponent(cache, CacheNotifier.class, mockNotifier, true);
+      mockSimpleNotifier = mock(CacheNotifier.class);
+      origSimpleNotifier = TestingUtil.replaceComponent(simpleCache, CacheNotifier.class, mockSimpleNotifier, true);
```
Wouldn't it be better to write a separate test for `SimpleCacheImpl`?

danberindei commented on documentation/src/main/asciidoc/user_guide/chapter-76-Simple_Cache.adoc:
```diff
@@ -0,0 +1,47 @@
+== Simple Cache
+
+Traditional local caches use the same architecture as clustered caches - interceptor stack. That way a lot of the functionality can be reused. However, if the advanced features are not needed and the performance is more important, the interceptor stack can be stripped away and simple cache can be used.
```
s/functionality can be reused/implementation can be reused/ ?

Also, for easy maintenance it's recommended to wrap lines and to start each sentence on a new line.

rvansa replied:
1) ok
2) really? Opening documentation, it is usually whole paragraph on one line. And I prefer using word wrapping in IDE (IDEA does that quite well once you map the to key shortcut) rather than rewrapping that manually when I need to edit that.

danberindei replied:
Yeah, it's usually one paragraph per line, but that' because of the way it got converted to asciidoc. Putting each sentence on a separate line makes it much easier to review changes in a PR.

danberindei commented on documentation/src/main/asciidoc/user_guide/chapter-76-Simple_Cache.adoc:
```diff
@@ -0,0 +1,47 @@
+
+* adding custom interceptors
+* Map Reduce Framework
+* Distributed Executors Framework
```
I wouldn't call the distributed executor a framework. Maybe the name fits a little better for Map/Reduce, because the user has to write multiple functions that work together, but I'd say it's a stretch there as well.

rvansa replied:
I did not come up with that name: http://infinispan.org/docs/7.2.x/user_guide/user_guide.html#_distributed_execution_framework

danberindei replied:
I stand corrected :)

danberindei commented on commons/src/main/java/org/infinispan/commons/equivalence/EntryEquivalence.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class EntryEquivalence<K, V> implements Equivalence<Map.Entry<K, V>> {
```
I don't think this is necessary, considering that `Map.entrySet()` isn't supposed to let the user add elements.

rvansa replied:
This is used in `EntrySet.retainAll()`, too.

danberindei replied:
I'm not sure it's needed there either, AFAICT all our other `retainAll()` implementations assume that `c.contains(e)` uses the same Equivalence and don't try to copy the input in another set with the cache's Equivalence.

danberindei replied:
I still think the entry equivalence isn't necessary... the `Collection.retainAll()` javadoc says "Retains only the elements in this collection that are contained in the specified collection", so it's defined based on the collection parameter's `contains()`, not on our `contains()`.

rvansa commented:
Addressed review comments, I'll rebase after we get feedback from @tristantarrant about the way how this should be configured.

tristantarrant commented:
The main question is: do we really need to expose this via XML yet ? Declarative configuration involves a ton of extra work (especially for server) which I don't think is worth it atm. I second @danberindei 's sentiment that the ability to inject an optimized interceptor stack on the fly would render this additional config redundant.

rvansa commented:
OK, I'll try if I encounter any major trouble with the hot switch. It will require one more volatile read and interface megamorphic call on each invocation, but if that's the price...

tristantarrant commented:
Summarizing my ideas from IRC and the above in clearer form:
I do not like the introduction of a new cache type for this exposed at the configuration level, as I'd rather have the code implicitly choose the best internal codepaths based on the features I select. For example, enabling security, returns a SecureCacheImpl. 
Although it would be nice to have extra-strict validation at the schema level so that a user cannot enable incompatible configuration items, I don't think it's worth the effort.

rvansa commented:
@danberindei @tristantarrant @wburns Rebased and squashed changes. Attribute simple-cache="true" is used to enable the cache now, rather than separate cache mode.
I've also fixed keySet(), values(), cacheEntrySet() and entrySet() methods to implement CacheXXX with the stream(), parallelstream() and spliterator() methods.
Could someone clear the labels?

gustavonalle commented:
needs rebasing and there's a compilation error on ```core```

rvansa commented:
@gustavonalle Thanks, rebased.

danberindei commented:
@rvansa there's still a compilation error in `SimpleCacheImpl`: http://ci.infinispan.org/viewLog.html?buildId=29715&buildTypeId=bt9&guest=1

rvansa commented:
Does CI do automatic merge and run the tests against merged build? I thought it runs tests on my branch.
Will added more methods today, so I need to rebase once again...

rvansa commented:
Let's see what CI says now... I would love to get the CI results to my mail, not having to peek on the PR.

danberindei commented:
CI runs the build on the merge, specifically to prevent integrating PRs that need rebasing :)

rvansa commented:
@danberindei Passes the CI, I hope you can find some time for this before I have to rebase again :)

danberindei commented:
@rvansa Sorry, needs another rebase: the documentation chapter number has to change.

danberindei commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -0,0 +1,1793 @@
+      putForExternalReadInternal(key, value, metadata, isCreatedRef);
+   }
+
+   protected void putForExternalReadInternal(K key, V value, Metadata metadata, ByRef.Boolean isCreatedRef) {
```
I think the two `putForExternalReadInternal` methods could be merged, or at least one of them should have its name changed.

danberindei commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -0,0 +1,1793 @@
+   }
+
+   @Override
+   public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
```
@wburns maybe we should support this in `CacheImpl` as well?

wburns replied:
We support it by default method, but it is pretty stupid (in this case it does a replace or putIfAbsent I believe).  Unfortunately Java 7/8 added a bunch of new methods which are prime for doing in a distributed/atomic way, but we haven't really made it a priority.

rvansa commented:
@danberindei Rebased & removed the EntryEquivalence. The 2 test failures are not related.

rvansa commented:
Any other changes needed?

danberindei commented:
Nope, integrated! Thanks Radim!
